### PR TITLE
Feat:Convert action bar as a component

### PR DESCRIPTION
**Pull request by manu-kr**
#### Please provide a link to the associated issue.

https://github.com/SAP/fundamental-ngx/issues/1263

#### Please provide a brief summary of this pull request.

This PR contains a code to enhance the action bar as a component.
This enhancement contains, providing an option to edit the Action bar Title and also displaying of overflow dynamically if the user configure more than 3 actions.

#### If this is a new feature, have you updated the documentation?

New Feature.

Documentation is pending.

### CODE REVIEW

**Comment by manu-kr** [comment]
@fkolar and @KevinOkamoto 

Below is the breif  design for the action bar for your refernce, I will update complete design details in wiki.

 <fdp-action-bar [actionbarTitle]="'Page Action bar component'"[showBackButton]=true [actionbarDescription]="'Action bar description from the component'">
      <fdp-action-bar-actions [actionItems]="actionItems" [placement]="'bottom-end'"></fdp-action-bar-actions>
    </fdp-action-bar>


and actionItems=actionItems= [{
      label: 'Save', // action label
      type:'main', // button type
      priority:1, // option to place the action in order
      callback: () => { // call back function
        alert('Save')
      }
    }, {
      label: 'Cancel',
      type:'primary',
      priority:2,
      callback: () => {
        alert('Cancel')
      }
    },
    {
      label: 'Rename',
      type:'main',
      priority:3,
      callback: () => {
        alert('Rename')
      }},
      {
        label: 'Delete',
        type:'primary',
        priority:3,
        callback: () => {
          alert('Delete')
        }
      },
      {
        label: 'Send',
        type:'main',
        priority:5,
        callback: () => {
          alert('Send')
        }}
 
    ]

**Comment by KevinOkamoto** [comment]
Question - can any of the action buttons be dropdown menus? If so I don't think this design allows for that.

File: libs/platform/src/lib/components/action-bar/action-bar-actions/action-bar-actions.component.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    console.log('button', this.buttonItems);
+    console.log('menu', this.menuItems);
+  }
+
+  onItemClick(item: ActionItem): void {
+
+    this.itemClick.emit(item);
+    this.cd.detectChanges();
+    item.callback();
+    console.log('action clicked', item.label);
```
[the remark this answers is not in the document]
Comment (manu-kr) [reply libs/platform/src/lib/components/action-bar/action-bar-actions/action-bar-actions.component.ts]: Removed all the console.log statements

File: libs/platform/src/lib/components/action-bar/action-bar-header/action-bar-header.component.html [hunk 2]
```diff
@@ -0,0 +1,31 @@
+<div #actionbar class="fd-action-bar action-bar-display">
+
+  <div class="fd-action-bar__back" *ngIf="showBackButton" (click)="backButtonClick.emit($event)">
+    <button id="actionbar__back-button" class="fd-button--light fd-button--compact sap-icon--nav-back"
```
[the remark this answers is not in the document]
Comment (manu-kr) [reply libs/platform/src/lib/components/action-bar/action-bar-header/action-bar-header.component.html]: Removed id="actionbar__back-button"

File: libs/platform/src/lib/components/action-bar/action-bar-header/action-bar-header.component.html [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  </ng-container>
+  <div class="fd-action-bar__actions">
+    <ng-content select="fdp-action-bar-actions" data-tag="fdp-action-bar-actions">
+    </ng-content>
+  </div>
+</div>
+
+<ng-template #actionbarContent>
+  <div class="fd-action-bar__header">
+    <h3 class="fd-action-bar__title" id="actionbar__title" *ngIf="!editMode" [attr.title]="actionbarTitle"
```
[the remark this answers is not in the document]
Comment (manu-kr) [reply libs/platform/src/lib/components/action-bar/action-bar-header/action-bar-header.component.html]: Removed id="actionbar__title"

File: libs/platform/src/lib/components/action-bar/action-bar-header/action-bar-header.component.html [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      data-tag="actionbar__title">
+      {{actionbarTitle}}</h3>
```
[the remark this answers is not in the document]
Comment (manu-kr) [reply libs/platform/src/lib/components/action-bar/action-bar-header/action-bar-header.component.html]: renamed to actionBarTitle

File: libs/platform/src/lib/components/action-bar/action-bar-header/action-bar-header.component.html [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    <div fd-form-set *ngIf="editMode">
+      <div fd-form-item>
+        <input id="actionbar__input" *ngIf="editMode" class="fd-input edit-actionbar-title" fd-form-control
```
[the remark this answers is not in the document]
Comment (manu-kr) [reply libs/platform/src/lib/components/action-bar/action-bar-header/action-bar-header.component.html]: removed id="actionbar__input"

File: libs/platform/src/lib/components/action-bar/action-bar-header/action-bar-header.component.html [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+          (focusout)="onFocusOut()" (keydown.enter)="$event.target.blur()" [(ngModel)]="actionbarTitle"
+          aria-required="true" data-tag="actionbar__input" />
+      </div>
+    </div>
+    <p id="actionbar__description" class="fd-action-bar__description" [attr.title]="actionbarDescription"
```
[the remark this answers is not in the document]
Comment (manu-kr) [reply libs/platform/src/lib/components/action-bar/action-bar-header/action-bar-header.component.html]: removed id="actionbar__description"

File: libs/platform/src/lib/components/action-bar/action-bar-header/action-bar-header.component.html [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      data-tag="actionbar__description">
+      {{actionbarDescription}}</p>
```
[the remark this answers is not in the document]
Comment (manu-kr) [reply libs/platform/src/lib/components/action-bar/action-bar-header/action-bar-header.component.html]: Renamed to actionBarDescription

**Comment by fkolar** [comment]
just use `[title], `[description]` you dont need to use "actionbar title" its implicit as the component is action bar

File: libs/platform/src/lib/components/action-bar/action-bar-header/action-bar-header.component.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  /**
+   * Emitted event when "back" button is clicked.
+   */
+  @Output() backButtonClick: EventEmitter<void> = new EventEmitter();
+
+  /**
+   * Emitted event when input textbox out of focus.
+   */
+
+  @Output() onRenameTitle: EventEmitter<string> = new EventEmitter<string>();
```
[the remark this answers is not in the document]
Comment (manu-kr) [reply libs/platform/src/lib/components/action-bar/action-bar-header/action-bar-header.component.ts]: chnaged to titleRenamed

**Comment by manu-kr** [comment]
> just use `[title], `[description]` you dont need to use "actionbar title" its implicit as the component is action bar
> 
> In terms of renaming @KevinOkamoto is right for the semantics but we need to also consider not tobe too descriprive when its are obvious.



> just use `[title], `[description]` you dont need to use "actionbar title" its implicit as the component is action bar
> 
> In terms of renaming @KevinOkamoto is right for the semantics but we need to also consider not tobe too descriprive when its are obvious.

> Okay Frank, I have renamed to [title] and [description]

**Comment by fkolar** [comment]
- you are not exporting library classes 
	- Create just like fundamentals public_api.ts for each module export ther eclass you want to be visible outside
	- and import this public_api.ts to mian index.ts

- There is confusion in naming, you have `fdp-action-bar`  inside  `ActionBarHeaderComponent`


**ActionBarHeaderComponent**

-> This should have depenency to our BUTTON, if we add some behavior to it, then we will be changing code on several places as somewhere we will be using <dfp-button and somewhere <button
  - Since button is not ready lets create new issue so we dont forget - we dont want to use 2 implementations 

-> Please use proper formating, you dont need new line in the actionbarheader constuctor
```

  constructor(
    public cd: ChangeDetectorRef) {
  }

```

-> please consider using `markForCheck()`, as you are using `onPush` strategy

-> where are you using <div #actionbar?  dont use this unless you want to reference it. Angular is making references and tries to bind it and its not nessesary in this case. 

-> what kind of construction is this?
	`  @Input() actionItems: [any];`

		- if you want to say its array of action items then any[]
		- but since we have created a action type we dont want to use any

		`actionItems: ActionItem[]` or `actionItems: Array<ActionItem>`



**ActionItem**
 -  using   `callback(): void` with such signature is not really effective you shoudl at minimum pass the event

 - butter solution is to use subject and be more reactive
 - More powerfulll solution is to use Subject 
 	`invoke: Subject<void>;`
   - Thsi will allow you to be modular and chain one action to another
    so the in the class you will be able to make something like this 

    ```ts
    invoke: this.doAction$
    ```

   where your subject can pipe this to do something else. 

   ```ts
    this.doAction$.pipe( 
    ```
   rxjs is really powerfull, I would just avoid using simple callback as it has not value. 


All CSS, TS
  - You proper formating. IDE usually can help you. 

e.g.

```
  .action-button-margin{
    margin-left: 8px;
}

```

missing space after margin. We will be turning more restrictive lint rules and all this would report and it would be too much to update

Always when you have file run some auto formating feature from your IDE.

the same applies for the `ActionModule`, 


**ActionBarActionsComponent**
	- you have unused import 
	- declaration is above constructors (we will turn our lint rules for this)
	MAX_BUTTONS = 3; should be a constant probably on top of the class or final readonly but better consts
	dont understand this syntax: `orderedActionItems[any]` - like above

	- like above `detectChanges` vs `markForCheck`

	- dont hardcode label `(click)="action.label=='Rename'?onRename` - what about i18n
	- again use formating for HTML template

**Comment by manu-kr** [comment]
> * you are not exporting library classes
>   
>   * Create just like fundamentals public_api.ts for each module export ther eclass you want to be visible outside
>   * and import this public_api.ts to mian index.ts
> * There is confusion in naming, you have `fdp-action-bar`  inside  `ActionBarHeaderComponent`
> 
> **ActionBarHeaderComponent**
> 
> -> This should have depenency to our BUTTON, if we add some behavior to it, then we will be changing code on several places as somewhere we will be using <dfp-button and somewhere <button
> 
> * Since button is not ready lets create new issue so we dont forget - we dont want to use 2 implementations
> 
> -> Please use proper formating, you dont need new line in the actionbarheader constuctor
> 
> ```
> 
>   constructor(
>     public cd: ChangeDetectorRef) {
>   }
> ```
> 
> -> please consider using `markForCheck()`, as you are using `onPush` strategy
> 
> -> where are you using <div #actionbar? dont use this unless you want to reference it. Angular is making references and tries to bind it and its not nessesary in this case.
> 
> -> what kind of construction is this?
> ` @Input() actionItems: [any];`
> 
> ```
> 	- if you want to say its array of action items then any[]
> 	- but since we have created a action type we dont want to use any
> 
> 	`actionItems: ActionItem[]` or `actionItems: Array<ActionItem>`
> ```
> 
> **ActionItem**
> 
> * using   `callback(): void` with such signature is not really effective you shoudl at minimum pass the event
> * butter solution is to use subject and be more reactive
> * More powerfulll solution is to use Subject
>   `invoke: Subject<void>;`
>   
>   * Thsi will allow you to be modular and chain one action to another
>     so the in the class you will be able to make something like this
>   
>   ```ts
>   invoke: this.doAction$
>   ```
>   
>   
>   where your subject can pipe this to do something else.
>   ```ts
>    this.doAction$.pipe( 
>   ```
>   
>   
>   rxjs is really powerfull, I would just avoid using simple callback as it has not value.
> 
> All CSS, TS
> 
> * You proper formating. IDE usually can help you.
> 
> e.g.
> 
> ```
>   .action-button-margin{
>     margin-left: 8px;
> }
> ```
> 
> missing space after margin. We will be turning more restrictive lint rules and all this would report and it would be too much to update
> 
> Always when you have file run some auto formating feature from your IDE.
> 
> the same applies for the `ActionModule`,
> 
> **ActionBarActionsComponent**
> - you have unused import
> - declaration is above constructors (we will turn our lint rules for this)
> MAX_BUTTONS = 3; should be a constant probably on top of the class or final readonly but better consts
> dont understand this syntax: `orderedActionItems[any]` - like above
> 
> ```
> - like above `detectChanges` vs `markForCheck`
> 
> - dont hardcode label `(click)="action.label=='Rename'?onRename` - what about i18n
> - again use formating for HTML template
> ```

@fkolar I have fixed the review comments , you can take a look.

**Comment by fkolar** [comment]
Manu please try to follow proper signature for commits

**Comment by manu-kr** [comment]
> Manu please try to follow proper signature for commits

@fkolar Sorry Frank. I will do that from now on.

**Comment by fkolar** [comment]
**ActionBarComponent**

- When you say 

`    @Input() editMode: boolean = false;` I would expect to see some other modes. I would just called it 

  ` @Input() editing: boolean = false;`

- change `constructor(public cd: ChangeDetectorRef) {}` to private unless you expect to expose this outside 

-Placement - do we have predefined list of placements? if Yes let;s create a type out of it 

- `@Input() displayOnlyMenu = false;`, when you are already using `showXXX` in the use also the same pattern here showXXX, hideXXX
  - What does this displayOnlyMenu really do? it switching between buttons actions and menu? 

- Let;s avoid using ` !important;`


**ActionBarActionsComponent**
- Please use IDE for formatting, I see some inconsistences that I commented before, contructor goes have variable declaration, you have unused imports. +. you have some unused properties. 

- You have some injected services with public modifier but you are not using them outside

- Please keep some consistency in formating somethings its hard to read. 
   * we shoud have some lines betweek each property, and we shoudl have some uniformity how to declare them. 


```
@Input() xxx: AAAA;

// or on the new line

@Input() 
xxx: AAAA;
```

When you start using one dont use other. I personally use on the new line as its easier to ready and it make some spaces 

- When you use directive try to use them as first thing in the element. The sape applies to` ngIf1` 

```
  <button
        class="action-button-margin fdp-action-button"
        [attr.title]="action.label"
        *ngFor="let action of buttonItems"
        fd-button
```
- Why do you need to use timeout here

```
focusFirst() {
        setTimeout(() => this.menulist.first.focus(), 0);
    }
```

- What if menulist is empty / undefined -> you get an error and app crash

**Comment by fkolar** [comment]
ActionBarComponent:

- Sync with latest on platform, always rebase before submiting your PR
   - your branch is out of sync
- Put some spaces between imports  and interface declaration  
- Why we are not using action-bar from fundamentals in the first place? 
	- the reason I am asking they upgraded all the component to have self-contained styles (please checkout action bar implementation in core) to support styles 0.3.0. 	Whole style is now part of action bar - they import scss.
- You need to align with this. 



ActionBarActionsComponent
 - you havent touched my previous comments, contructor cannot be as first thing you have property declaration, then you have contructor. 
    - It cannot look right to you as well when the action bar component is structured differently. 

- It seems like you havent addressed most of the comments for this component. 


Such as: 
  -> When you use directive try to use them as first thing in the element. The sape applies to ngIf
  -> Why do you need to use timeout here...
  -> What if menulist is empty / undefined -> you get an error and app crash


Before you push next commit please squash the commits locally , there are too many. You are publishing feature so you dont want to say 

feat(xxx): Fix for review... 
 - when you fixing something then its not reallly a feat right ? 

Make sure you have here only 1 commit e.g.: `feat(platform): Convert action bar as a component `
adn squash the rest of hte commit into this one. ping me on slack if any questions.

**Comment by fkolar** [comment]
https://github.com/SAP/fundamental-ngx/wiki/wip:-Contribute

**Comment by manu-kr** [comment]
> ActionBarComponent:
> 
> * Sync with latest on platform, always rebase before submiting your PR
>   
>   * your branch is out of sync
> * Put some spaces between imports  and interface declaration
> * Why we are not using action-bar from fundamentals in the first place?
>   
>   * the reason I am asking they upgraded all the component to have self-contained styles (please checkout action bar implementation in core) to support styles 0.3.0. 	Whole style is now part of action bar - they import scss.
> * You need to align with this.
> 
> ActionBarActionsComponent
> 
> * you havent touched my previous comments, contructor cannot be as first thing you have property declaration, then you have contructor.
>   
>   * It cannot look right to you as well when the action bar component is structured differently.
> * It seems like you havent addressed most of the comments for this component.
> 
> Such as:
> -> When you use directive try to use them as first thing in the element. The sape applies to ngIf
> -> Why do you need to use timeout here...
> -> What if menulist is empty / undefined -> you get an error and app crash
> 
> Before you push next commit please squash the commits locally , there are too many. You are publishing feature so you dont want to say
> 
> feat(xxx): Fix for review...
> 
> * when you fixing something then its not reallly a feat right ?
> 
> Make sure you have here only 1 commit e.g.: `feat(platform): Convert action bar as a component `
> adn squash the rest of hte commit into this one. ping me on slack if any questions.

@fkolar 

Q1


> ActionBarComponent:
> 
> * Sync with latest on platform, always rebase before submiting your PR
>   
>   * your branch is out of sync
    Ans:  Rebased from master and merged latest changes 
> * Put some spaces between imports  and interface declaration
   Ans: Done
> * Why we are not using action-bar from fundamentals in the first place?
Ans :    action-bar from fundamental-ngx is a directives and i am reusing all those directives in platform action bar component.
>   * the reason I am asking they upgraded all the component to have self-contained styles (please checkout action bar implementation in core) to support styles 0.3.0. 	Whole style is now part of action bar - they import scss.
Ans: I have verified with the styles 0.3.0
> * You need to align with this.
> 
> ActionBarActionsComponent
> 
> * you havent touched my previous comments, contructor cannot be as first thing you have property declaration, then you have constructor.
>Ans:   I addressed all this issue in my previous commit itself, I am not sure why you were not able to see.
>   * It cannot look right to you as well when the action bar component is structured differently.
> * It seems like you havent addressed most of the comments for this component.
> 
> Such as:
> -> When you use directive try to use them as first thing in the element. The sape applies to ngIf
Ans: i have modified this.
> -> Why do you need to use timeout here...
Ans: When user click on overflow icon, First item in the menu should get focus, If didn't use timeout it won't get focus.
> -> What if menulist is empty / undefined -> you get an error and app crash
> Ans: I am checking the empty condition before that if menulist is empty it won't display overflow icon itself.
 <button
                *ngIf="menuItems?.length > 0"...> 


> 
> Before you push next commit please squash the commits locally , there are too many. You are publishing feature so you dont want to say


> 
> feat(xxx): Fix for review...
> 
> * when you fixing something then its not reallly a feat right ?
> 
> Make sure you have here only 1 commit e.g.: `feat(platform): Convert action bar as a component `
> adn squash the rest of hte commit into this one. ping me on slack if any questions.

**Comment by fkolar** [comment]
* When we use directive let’s try to use them as first thing before any other attribute. What I meant was  instead of using this directive in the middle:

```html
 <input
                    *ngIf="editing"
                    class="fd-input edit-actionbar-title"
                    fd-form-control
                    (focusout)="onFocusOut()"
                    (keydown.enter)="$event.target.blur()"
                    [(ngModel)]="title"
```

 you can do this:

```html
 <input   fd-form-control
                    *ngIf="editing"
                    class="fd-input edit-actionbar-title"       
```
so developer see immediately that this input is enhanced by directive

Otherwise it looks good !!

Last question for the `setTimeout` :
 
- What is happening since you set onPush CD strategy you need to take care of things by your self. 
- setting timeout is just another things to run change detection. Have you tried to call in this case `detectChanges`  ?

-

**Comment by manu-kr** [comment]
@fkolar 


> * When we use directive let’s try to use them as first thing before any other attribute. What I meant was  instead of using this directive in the middle:
> 
> ```
>  <input
>                     *ngIf="editing"
>                     class="fd-input edit-actionbar-title"
>                     fd-form-control
>                     (focusout)="onFocusOut()"
>                     (keydown.enter)="$event.target.blur()"
>                     [(ngModel)]="title"
> ```
> 
> you can do this:
> 
> ```
>  <input   fd-form-control
>                     *ngIf="editing"
>                     class="fd-input edit-actionbar-title"       
> ```
> 

>``` Ans: Done.
> so developer see immediately that this input is enhanced by directive. You have it all over the templates.
> 
> Otherwise it looks good !!
> 
> Last question for the `setTimeout` :
> 
> * What is happening since you set onPush CD strategy you need to take care of things by your self.
> * setting timeout is just another things to run change detection. Have you tried to call in this case `detectChanges`  ?

Ans: I have tried with this approach, but it doesn't help.

**Comment by manu-kr** [comment]
> @fkolar
> 
> > * When we use directive let’s try to use them as first thing before any other attribute. What I meant was  instead of using this directive in the middle:
> > 
> > ```
> >  <input
> >                     *ngIf="editing"
> >                     class="fd-input edit-actionbar-title"
> >                     fd-form-control
> >                     (focusout)="onFocusOut()"
> >                     (keydown.enter)="$event.target.blur()"
> >                     [(ngModel)]="title"
> > ```
> > 
> > 
> > you can do this:
> > ```
> >  <input   fd-form-control
> >                     *ngIf="editing"
> >                     class="fd-input edit-actionbar-title"       
> > ```
> 
> > ```   Ans: Done.
> > so developer see immediately that this input is enhanced by directive. You have it all over the templates.
> > 
> > Otherwise it looks good !!
> > 
> > Last question for the `setTimeout` :
> > 
> > * What is happening since you set onPush CD strategy you need to take care of things by your self.
> > * setting timeout is just another things to run change detection. Have you tried to call in this case `detectChanges`  ?
> > ```
> 
> Ans: I have tried with this approach, but it doesn't help.

**Comment by fkolar** [comment]
1000 changed files? I think you need to rebase properly so you can see only your changes on top and the rest should be identical with master so there should be no need to have that file in your commit. 

Dont rebase before you resolve this.

**Comment by droshev** [comment]
@manu-kr any comments?

**Comment by fkolar** [comment]
@manu-kr let's revisit this once more and lets start with
the correct rebase so there are only components that are changes and I see there are already conflicting files, 

or you can close this PR and submit another one to correct this. wiht 1000 files its imposible to review.

**Comment by manu-kr** [comment]
> @manu-kr let's revisit this once more and lets start with
> the correct rebase so there are only components that are changes and I see there are already conflicting files,
> 
> or you can close this PR and submit another one to correct this. wiht 1000 files its imposible to review.

@fkolar 
> This PR changes are already moved to different PR (https://github.com/SAP/fundamental-ngx/pull/1398) and its merged to master. so we can close this PR.

**Pull request closed without merging**